src/realm/object-store/c_api/query.cpp

```diff
@@ -125,30 +127,67 @@ struct QueryArgumentsAdapter : query_builder::Arguments {
         check_index(i);
         return m_args[i].type == RLM_TYPE_NULL;
     }
+    DataType type_for_argument(size_t i) override
+    {
+        check_index(i);
+        if (m_args[i].type == RLM_TYPE_INT) {
```

Comment:
What's wrong with `switch` statements all of a sudden? :-)

It would be good to get compiler warnings for uncovered cases.